File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/NativeIdentity.kt
package dev.idenqa.sdk

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyInfo
import android.security.keystore.KeyProperties
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.KeyStore
import java.security.MessageDigest
import java.security.PrivateKey
import java.security.Signature
import java.security.spec.ECGenParameterSpec
import java.time.Instant
import java.util.Base64

interface NativeProofKey {
    suspend fun publicKey(): ByteArray
    suspend fun sign(message: ByteArray): ByteArray
}

fun interface NativeAttestationProvider { suspend fun attestation(): String? }

class AndroidP256ProofKey(private val alias: String = "dev.idenqa.native-proof.v1") : NativeProofKey {
    init { if (alias.isBlank()) throw IdenqaException.InvalidConfiguration }

    override suspend fun publicKey(): ByteArray = pair().certificate.publicKey.encoded.copyOf()

    override suspend fun sign(message: ByteArray): ByteArray {
        val signature = Signature.getInstance("SHA256withECDSA")
        signature.initSign(pair().privateKey)
        signature.update(message)
        return signature.sign()
    }

    private fun pair(): KeyStore.PrivateKeyEntry {
        val store = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
        (store.getEntry(alias, null) as? KeyStore.PrivateKeyEntry)?.let {
            requireHardwareBacked(it.privateKey)
            return it
        }
        val generator = KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_EC, "AndroidKeyStore")
        generator.initialize(
            KeyGenParameterSpec.Builder(alias, KeyProperties.PURPOSE_SIGN)
                .setAlgorithmParameterSpec(ECGenParameterSpec("secp256r1"))
                .setDigests(KeyProperties.DIGEST_SHA256)
                .setUserAuthenticationRequired(false)
                .build(),
        )
        generator.generateKeyPair()
        return (store.getEntry(alias, null) as? KeyStore.PrivateKeyEntry)?.also { requireHardwareBacked(it.privateKey) }
            ?: throw IdenqaException.SecureStorage
    }

    private fun requireHardwareBacked(key: PrivateKey) {
        val factory = KeyFactory.getInstance(key.algorithm, "AndroidKeyStore")
        val info = factory.getKeySpec(key, KeyInfo::class.java)
        @Suppress("DEPRECATION")
        if (!info.isInsideSecureHardware) throw IdenqaException.HardwareSecurityUnavailable
    }
}

class NativeBootstrapIdentity(
    val applicationId: String,
    private val key: NativeProofKey,
    private val attestationProvider: NativeAttestationProvider? = null,
) {
    init {
        if (applicationId.isBlank() || applicationId.length > 255 || !applicationId.all { it.isLetterOrDigit() || it in "._-" }) {
            throw IdenqaException.InvalidConfiguration
        }
    }

    companion object {
        fun installedApplication(context: Context, key: NativeProofKey, attestationProvider: NativeAttestationProvider? = null) =
            NativeBootstrapIdentity(context.applicationContext.packageName, key, attestationProvider)
    }

    internal suspend fun request(token: String, capabilities: CapabilityAdvertisement, createdAt: Instant): NativeBootstrapRequest {
        val created = createdAt.epochSecond
        if (created <= 0) throw IdenqaException.InvalidConfiguration
        val tuple = capabilities.platform + "\u0000" + capabilities.sdkVersion + "\u0000" + capabilities.implementedMethods.sorted().joinToString("\u001f") + "\u0000" + capabilities.currentlyAvailableMethods.sorted().joinToString("\u001f")
        val canonical = "idq-native-bootstrap\u0000v1\u0000$applicationId\u0000POST\u0000/v1/capture/native/bootstrap\u0000$created\u0000${sha256(token.toByteArray())}\u0000${sha256(tuple.toByteArray())}"
        return NativeBootstrapRequest(
            applicationId, key.publicKey().base64Url(), created, "ES256", "der",
            key.sign(canonical.toByteArray()).base64Url(), attestationProvider?.attestation(), capabilities,
        )
    }

    private fun sha256(value: ByteArray) = MessageDigest.getInstance("SHA-256").digest(value).joinToString("") { "%02x".format(it) }
    private fun ByteArray.base64Url() = Base64.getUrlEncoder().withoutPadding().encodeToString(this)
}

internal data class NativeBootstrapRequest(
    val applicationId: String,
    val proofKey: String,
    val proofCreatedAt: Long,
    val proofAlgorithm: String,
    val proofFormat: String,
    val proof: String,
    val attestation: String?,
    val capabilities: CapabilityAdvertisement,
)

File: sdk/swift/Sources/Idenqa/NativeIdentity.swift
import CryptoKit
import Foundation
#if canImport(Security)
import Security
#endif

public protocol NativeProofKey: Sendable {
    func publicKey() async throws -> Data
    func sign(_ message: Data) async throws -> Data
}

public protocol NativeAttestationProvider: Sendable {
    func attestation() async throws -> String?
}

public struct NativeBootstrapIdentity: Sendable {
    public let applicationID: String
    private let key: any NativeProofKey
    private let attestationProvider: (any NativeAttestationProvider)?

    public init(applicationID: String, key: any NativeProofKey, attestationProvider: (any NativeAttestationProvider)? = nil) throws {
        guard Self.validApplicationID(applicationID) else { throw IdenqaError.invalidConfiguration }
        self.applicationID = applicationID
        self.key = key
        self.attestationProvider = attestationProvider
    }

    public static func installedApplication(key: any NativeProofKey, attestationProvider: (any NativeAttestationProvider)? = nil) throws -> Self {
        guard let identifier = Bundle.main.bundleIdentifier else { throw IdenqaError.invalidConfiguration }
        return try Self(applicationID: identifier, key: key, attestationProvider: attestationProvider)
    }

    func request(token: String, capabilities: CapabilityAdvertisement, createdAt: Date) async throws -> NativeBootstrapRequest {
        let created = Int64(createdAt.timeIntervalSince1970)
        guard created > 0 else { throw IdenqaError.invalidConfiguration }
        let tokenDigest = SHA256.hash(data: Data(token.utf8)).hex
        let tuple = capabilities.platform + "\0" + capabilities.sdkVersion + "\0" + capabilities.implementedMethods.sorted().joined(separator: "\u{1f}") + "\0" + capabilities.currentlyAvailableMethods.sorted().joined(separator: "\u{1f}")
        let bodyDigest = SHA256.hash(data: Data(tuple.utf8)).hex
        let canonical = "idq-native-bootstrap\0v1\0\(applicationID)\0POST\0/v1/capture/native/bootstrap\0\(created)\0\(tokenDigest)\0\(bodyDigest)"
        let signature = try await key.sign(Data(canonical.utf8))
        let publicKey = try await key.publicKey()
        return NativeBootstrapRequest(
            applicationID: applicationID,
            proofKey: publicKey.base64URL,
            proofCreatedAt: created,
            proofAlgorithm: "ES256",
            proofFormat: "der",
            proof: signature.base64URL,
            attestation: try await attestationProvider?.attestation(),
            capabilities: capabilities
        )
    }

    private static func validApplicationID(_ value: String) -> Bool {
        !value.isEmpty && value.utf8.count <= 255 && value.allSatisfy { $0.isASCII && ($0.isLetter || $0.isNumber || "._-".contains($0)) }
    }
}

struct NativeBootstrapRequest: Codable, Sendable {
    let applicationID: String
    let proofKey: String
    let proofCreatedAt: Int64
    let proofAlgorithm: String
    let proofFormat: String
    let proof: String
    let attestation: String?
    let capabilities: CapabilityAdvertisement

    enum CodingKeys: String, CodingKey {
        case applicationID = "application_id"
        case proofKey = "proof_key"
        case proofCreatedAt = "proof_created_at"
        case proofAlgorithm = "proof_algorithm"
        case proofFormat = "proof_format"
        case proof, attestation, capabilities
    }
}

#if canImport(Security) && canImport(CryptoKit)
public actor SecureEnclaveP256ProofKey: NativeProofKey {
    private let tag: String
    private var key: SecureEnclave.P256.Signing.PrivateKey?

    public init(tag: String = "dev.idenqa.native-proof.v1") throws {
        guard !tag.isEmpty, SecureEnclave.isAvailable else { throw IdenqaError.hardwareSecurityUnavailable }
        self.tag = tag
    }

    public func publicKey() throws -> Data { try load().publicKey.x963Representation }

    public func sign(_ message: Data) throws -> Data {
        try load().signature(for: message).derRepresentation
    }

    private func load() throws -> SecureEnclave.P256.Signing.PrivateKey {
        if let key { return key }
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: tag,
            kSecAttrAccount as String: "p256-signing-key",
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        let loaded: SecureEnclave.P256.Signing.PrivateKey
        if status == errSecSuccess, let data = result as? Data {
            loaded = try SecureEnclave.P256.Signing.PrivateKey(dataRepresentation: data)
        } else if status == errSecItemNotFound {
            loaded = try SecureEnclave.P256.Signing.PrivateKey()
            var item = query
            item.removeValue(forKey: kSecReturnData as String)
            item.removeValue(forKey: kSecMatchLimit as String)
            item[kSecValueData as String] = loaded.dataRepresentation
            item[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
            guard SecItemAdd(item as CFDictionary, nil) == errSecSuccess else { throw IdenqaError.secureStorage }
        } else {
            throw IdenqaError.secureStorage
        }
        key = loaded
        return loaded
    }
}
#endif

private extension Data {
    var base64URL: String { base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "") }
}

private extension SHA256.Digest {
    var hex: String { map { String(format: "%02x", $0) }.joined() }
}
